File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root:
            return None
        
        if root == p or root == q:
            return root
        
        left_ST = self.lowestCommonAncestor(root.left, p, q)
        right_ST = self.lowestCommonAncestor(root.right, p, q)
        
        if left_ST and right_ST:
            return root
        else:
            return left_ST or right_ST

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    
    // leaf node edge case needed
    if (!root) return null
    
    // if either p or q are not null, then this is the LCA
    if (root === p || root === q) return root
        
    let leftSt
    let rightSt
    
    leftSt = lowestCommonAncestor(root.left, p , q)
    rightSt = lowestCommonAncestor(root.right, p , q)
        
    if (leftSt && rightSt) {
        return root
    }
    else return leftSt || rightSt
};